File: app/src/main/java/com/anadi/weatherinfo/details/DetailsContract.java
package com.anadi.weatherinfo.details;

import androidx.annotation.StringRes;

import com.anadi.weatherinfo.repository.data.WeatherInfo;

import java.util.Observer;

public interface DetailsContract {

    interface View {

        void loading();

        void onError(@StringRes int resId);

        void onUpdateSuccess();
    }

    interface Presenter extends Observer {
        void update();

        void subscribe();

        void unsubscribe();

        WeatherInfo getInfo(int id);
    }

    interface Model {
        WeatherInfo getInfo(int id);

        boolean update(int id);

        boolean needUpdate(int id);
    }
}

File: app/src/main/java/com/anadi/weatherinfo/details/DetailsContract.kt
package com.anadi.weatherinfo.details

import androidx.annotation.StringRes
import com.anadi.weatherinfo.repository.data.WeatherInfo
import java.util.*

interface DetailsContract {
    interface View {
        fun loading()
        fun onError(@StringRes resId: Int)
        fun onUpdateSuccess()
    }

    interface Presenter : Observer {
        fun update()
        fun subscribe()
        fun unsubscribe()
        fun getInfo(id: Int): WeatherInfo?
    }

    interface Model {
        fun getInfo(id: Int): WeatherInfo?
        fun update(id: Int): Boolean
        fun needUpdate(id: Int): Boolean
    }
}